### PR TITLE
amount parses 0 msats

### LINKED ISSUE #1529: Amount is not parsing 0 msat
0 msat is representable as sats - just 0, so we should allow it.

### PR DESCRIPTION
**Pull request by td-bn**
Closes #1529 

0 msats should be parasable.

### CODE REVIEW

**Comment by td-bn** [comment]
Is there a better way to parse 0 from string?

**Comment by Kixunil** [comment]
Actually, I just realized we should accept any multiple of 1000 for msat and there's no test for this and I think it doesn't parse now. Would you like to fix it?

**Comment by Kixunil** [comment]
Oh, WTF, looks like I don't understand the code. But I guess it's OK if the tests pass. :)

**Review by Kixunil — DISMISSED** [review]
ACK 472e0da3b24aab4b39663dfb9afd8fc4530f4cce

File: bitcoin/src/amount.rs [hunk 1]
```diff
@@ -1530,6 +1533,16 @@ mod tests {
...
+    
+    #[test]
+    fn from_str_zero() {
+        let denoms = vec!["BTC", "mBTC", "uBTC", "nBTC", "pBTC", "bits", "sats", "msats"];
+        for denom in denoms {
+            let s = format!("0 {}", denom);
+            assert!(Amount::from_str(&s).is_ok(), "Failed to create amount from {}", s);
+        }
+    }
+
```
Comment (tcharding) [inline bitcoin/src/amount.rs:1545]: Oh I didn't realise we were supposed to support strange things like "-0" and "000", this test fails
```rust
    #[test]
    fn from_str_zero() {
        let denoms = vec!["BTC", "mBTC", "uBTC", "nBTC", "pBTC", "bits", "sats", "msats"];
        for denom in denoms {
            for v in vec!["0", "-0", "000"] {
                let s = format!("{} {}", v, denom);
                match Amount::from_str(&s) {
                    Err(e) => panic!("Failed to crate amount from {}: {:?}", s, e),
                    Ok(amount) => assert_eq!(amount, Amount::from_sat(0)),
                }
            }
        }
    }
```

File: bitcoin/src/amount.rs [hunk 2]
```diff
@@ -1924,14 +1937,23 @@ mod tests {
...
-        case("000 msat", Err(E::TooPrecise));
-        case("-000 msat", Err(E::TooPrecise));
+        scase("0 msat", Ok(SignedAmount::from_sat(0)));
+        scase("-0 msat", Ok(SignedAmount::from_sat(0)));
+        scase("000 msat", Ok(SignedAmount::from_sat(0)));
+        scase("-000 msat", Ok(SignedAmount::from_sat(0)));
+        case("0 msat", Ok(Amount::from_sat(0)));
+        case("-0 msat", Ok(Amount::from_sat(0)));
+        case("000 msat", Ok(Amount::from_sat(0)));
+        case("-000 msat", Ok(Amount::from_sat(0)));
```
Comment (tcharding) [inline bitcoin/src/amount.rs:1947]: Perhaps just remove all this stuff if you use the failing test I posted.

File: bitcoin/src/amount.rs (hunk 1 above)
Comment (Kixunil) [reply bitcoin/src/amount.rs:1545]: You love vecs, don't you? ;)

**Comment by Kixunil** [comment]
It's an interesting question whether we should allow `-0` for *not signed* `Amount`. Mathematically it makes sense, but forbidding it could catch things like type being used (if one uses minus sign on zero probably because there's something silly like `format!("-{}", x)`)

FTR, [`core` doesn't allow `-0` for unsigned types](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a1d8f9b0fdcd7fbd1e2004f24f4c02bd).

**Comment by apoelstra** [comment]
Yeah, let's not allow -0 ... while is the same number as 0 when considered as a mathematical number, I'd argue that it's a different kind of object when considered in the platonic type system in my head.

File: bitcoin/src/amount.rs (hunk 1 above)
Comment (tcharding) [reply bitcoin/src/amount.rs:1555]: And Big-O n squared algorithms ;)

**Review by Kixunil — APPROVED** [review]
ACK 56de28da7a0c1f21fad0c93fa600bcb6699692be

File: bitcoin/src/amount.rs [hunk 3]
```diff
@@ -1544,6 +1547,30 @@ mod tests {
     use serde_test;
 
     use super::*;
+
```
Comment (tcharding) [inline bitcoin/src/amount.rs:1550]: There is trailing whitespace on this line. You can use the githooks to catch these (check out the README) for how to do that. If you are wondering how I knew, I have my shell and editor configured to highlight with red if there is trailing whitespace. I'll ack anyways, its no biggy - just something to know for next time. Cheers.

**Review by tcharding — APPROVED** [review]
ACK 56de28da7a0c1f21fad0c93fa600bcb6699692be

**Comment by tcharding** [comment]
The PR title has `WIP` in it still :)

**Review by Kixunil — APPROVED** [review]
ACK 0cf6da4f03be11ed303cda7fdb95538c491a3305

**Review by tcharding — APPROVED** [review]
ACK 0cf6da4f03be11ed303cda7fdb95538c491a3305

**Review by apoelstra — APPROVED** [review]
ACK 0cf6da4f03be11ed303cda7fdb95538c491a3305

**Pull request merged**